### ISSUE TITLE
ssh filesystem and put_files are incompatible

### ISSUE THREAD

**Issue by ghislainp** [fsspec/filesystem_spec#1071]
Using a ssh filesystem  and put_files gives an error

```python
import fsspec
import sshfs
fs = fsspec.filesystem("ssh", host='xx.xx.xx.xx', username='eouser') 
fs.put_file(lpath="t", rpath="/home/me/t")
```
I receive an exception:
```
File "/home/me/miniconda3/envs/olci/lib/python3.9/site-packages/fsspec/callbacks.py", line 66, in relative_update
    self.value += inc
TypeError: unsupported operand type(s) for +=: 'int' and 'NoneType'
```
The reason is because the 'write' method in paramiko does not return the length of written data.
https://github.com/jbouse-debian/paramiko/blob/master/paramiko/file.py#L385

Unless paramiko changes their function, the method 'put_file' in spec.py should test segment_len to be None and skip the update... or use len(data) as a default as follows:

```python
                    segment_len = f2.write(data)
                    if segment_len is None:
                          segment_len = len(data)
                    callback.relative_update(segment_len)
```

I'm not sure if this is a good solution...

**Comment by martindurant** [comment]
@efiop , time to ditch paramiko-based ssh?

@ghislainp , would you like to try with https://github.com/fsspec/sshfs ?

**Comment by ghislainp** [comment]
I'm using prefect which is using fsspec in a generic way...

**Comment by martindurant** [comment]
I suppose you could try temporarily with 
```
import fsspec
import sshfs
fsspec.register_implementation("ssh", sshfs.SSHFileSystem)
```
@efiop , @pmrowla : I note tha sshfs is not noarch (surprised me) and has not been released for osx-arm64.

**Comment by efiop** [comment]
@martindurant Hm, that is indeed surprising. We only have `any` wheel in https://pypi.org/project/sshfs/#files , which I thought meant noarch (same with underlying asyncssh). Are you sure it is not some other package that is getting compiled?  I'm also on m1, but don't see anything being compiled when I install sshfs.

**Comment by efiop** [comment]
@martindurant Ah, right 😅 I remember we tried `noarch` before, but maybe it wasn't with sshfs conda package specifically and so we copied it over from another place with no second thoughts. We'll double-check.

**Comment by efiop** [comment]
@martindurant Yeah, looks like we just copypasted it back then without trying `noarch` first. Fixed in https://github.com/conda-forge/sshfs-feedstock/pull/20 . Thanks for the heads up! 🙏

**Comment by ghislainp** [comment]
this does not work well with prefect for two reasons:

1- it seems that prefect uses the full uri for the destination path in 
put_file, I don't know if it is valid but anyway the SFTP implementation 
was able to remove the protocle and host part.|The SSHFileSystem 
implementation does not remove it and it fails. I don't have a direct 
way to change this in prefect.|

|2- |The SFTP|| implementation automatically create the required 
directory (e.g. uri ssh://host/directory/subdirectory/file.txt works 
even if directory does not exist yet). The |SSHFileSystem does not 
create these directories. Prefect relies on the first behavior.
|

Let me know if these problemes are considered as bugs in the 
SSHFileSystem or if I should report the issue to prefect.

**Comment by ghislainp** [comment]
> I suppose you could try temporarily with
> 
> ```
> import fsspec
> import sshfs
> fsspec.register_implementation("ssh", sshfs.SSHFileSystem)
> ```
> 

This does not work well with prefect for two reasons:

1- it seems that prefect uses the full uri for the destination path in put_file, I don't know if it is valid but anyway the SFTP implementation was able to remove the protocle and host part. The SSHFileSystem implementation does not remove it and it fails. I don't have a direct way to change this in prefect.

2- The SFTP implementation automatically create the required directory (e.g. uri ssh://host/directory/subdirectory/file.txt works even if directory does not exist yet). The SSHFileSystem does not create these directories. Prefect relies on the first behavior.

Let me know if these problemes are considered as bugs in the SSHFileSystem or if I should report the issue to prefect.

**Comment by efiop** [comment]
> 2- The SFTP implementation automatically create the required directory (e.g. uri ssh://host/directory/subdirectory/file.txt works even if directory does not exist yet). The SSHFileSystem does not create these directories. Prefect relies on the first behavior.

Prefect should probably be creating that directory explicitly before `put_file`. The behavior varies across different filesystems (usually simply because s3fs/gcsfs/adlfs don't have a knowtion of directories), but I would argue that not auto-creating directories is the correct default behaviour. That being said, it is worth noting that localfs has `auto_mkdir` option, which we could consider supporting in sshfs, but overall this seems like a bit of a hack.

**Comment by martindurant** [comment]
> I'm not sure if this is a good solution...

@ghislainp , I think your solution is a good one until we are certain we want to swap out the SSH implementation. Would you like to put it in a PR?
